File: src/main/scala/robosky/uplands/interop/AegisaltRecipeDisplay.scala
package robosky.uplands.interop

import java.util
import java.util.Optional

import com.google.common.collect.Lists
import me.shedaniel.rei.api.{EntryStack, RecipeDisplay}
import net.minecraft.inventory.Inventory
import net.minecraft.item.ItemStack
import net.minecraft.recipe.Recipe
import net.minecraft.util.Identifier
import robosky.uplands.UplandsMod
import robosky.uplands.block.machine.infuser.AegisaltRecipe
import robosky.uplands.item.ItemRegistry

import scala.collection.JavaConverters._

case class AegisaltRecipeDisplay(recipe: AegisaltRecipe) extends RecipeDisplay {
  private val fuel: util.ArrayList[EntryStack] = Lists.newArrayList(EntryStack.create(ItemRegistry.AEGISALT_CRYSTAL))
  private val input: util.List[util.List[EntryStack]] = Lists.newArrayList(recipe.ingredients
    .map(_.getMatchingStacksClient.toList.map(EntryStack.create).asJava).asJava)
  input.add(fuel)

  def getFuel: util.List[EntryStack] = fuel

  override def getOutputEntries: util.List[EntryStack] = Lists.newArrayList(EntryStack.create(recipe.getOutput))

  override def getRecipeCategory: Identifier = UplandsMod :/ "plugins/aegisalt_infusions"

  override def getRequiredEntries: util.List[util.List[EntryStack]] = getInputEntries

  override def getInputEntries: util.List[util.List[EntryStack]] = input
}

File: src/main/java/robosky/uplands/interop/AegisaltRecipeDisplay.java
package robosky.uplands.interop;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import me.shedaniel.rei.api.EntryStack;
import me.shedaniel.rei.api.RecipeDisplay;
import robosky.uplands.UplandsMod;
import robosky.uplands.block.machine.infuser.AegisaltRecipe;
import robosky.uplands.item.ItemRegistry;

import net.minecraft.item.ItemStack;
import net.minecraft.recipe.Ingredient;
import net.minecraft.util.Identifier;

public final class AegisaltRecipeDisplay implements RecipeDisplay {

    private final AegisaltRecipe recipe;
    private final List<EntryStack> fuel;
    private final List<List<EntryStack>> input;

    public AegisaltRecipeDisplay(AegisaltRecipe recipe) {
        this.recipe = recipe;
        this.fuel = Collections.singletonList(EntryStack.create(ItemRegistry.AEGISALT_CRYSTAL));
        List<List<EntryStack>> ls = new ArrayList<>();
        for(Ingredient ingredient : recipe.getIngredients()) {
            List<EntryStack> stacks = new ArrayList<>();
            for(ItemStack stack : ingredient.getMatchingStacksClient()) {
                stacks.add(EntryStack.create(stack));
            }
            ls.add(Collections.unmodifiableList(stacks));
        }
        ls.add(fuel);
        this.input = Collections.unmodifiableList(ls);
    }

    public List<EntryStack> getFuel() {
        return fuel;
    }

    @Override
    public List<EntryStack> getOutputEntries() {
        return Collections.singletonList(EntryStack.create(this.recipe.getOutput()));
    }

    @Override
    public Identifier getRecipeCategory() {
        return UplandsMod.id("plugins/aegisalt_infusions");
    }

    @Override
    public List<List<EntryStack>> getRequiredEntries() {
        return this.getInputEntries();
    }

    @Override
    public List<List<EntryStack>> getInputEntries() {
        return this.input;
    }
}
